components/brave_vpn/winvpntool.cc

```diff
@@ -0,0 +1,761 @@
+    }
+    // The first RASENTRYNAME structure in the array must contain the structure
+    // size
+    lp_ras_entry_name[0].dwSize = sizeof(RASENTRYNAME);
+
+    // Call RasEnumEntries to enumerate all RAS entry names
+    dw_ret = RasEnumEntries(NULL, NULL, lp_ras_entry_name, &dw_cb, &dw_entries);
+
+    // If successful, print the RAS entry names
+    if (ERROR_SUCCESS == dw_ret) {
+      wprintf(L"The following RAS entry names were found:\n");
+      for (DWORD i = 0; i < dw_entries; i++) {
+        wprintf(L"%s\n", lp_ras_entry_name[i].szEntryName);
+        dw_ret = PrintEntryDetails(lp_ras_entry_name[i].szEntryName);
+      }
+    }
+    // Deallocate memory for the connection buffer
+    HeapFree(GetProcessHeap(), 0, lp_ras_entry_name);
+    lp_ras_entry_name = NULL;
+    return ERROR_SUCCESS;
+  }
+
+  // There was either a problem with RAS or there are RAS entry names to
+  // enumerate
+  if (dw_entries >= 1) {
+    wprintf(L"The operation failed to acquire the buffer size.\n\n");
+  } else {
+    wprintf(L"There were no RAS entry names found:.\n\n");
+  }
+
+  return dw_ret;
+}
+
+void copy_dword_bytes(BYTE* bytes, DWORD value) {
+  if (bytes) {
+    union {
+      DWORD value;
+      BYTE bytes[4];
+    } converter;
+    converter.value = value;
+    memcpy(bytes, converter.bytes, 4);
+  }
+}
+
+// NOTE: This code is never called, but this is how the magic number set in
+// brave_vpn::CreateEntry() can be created for the `CustomIPSecPolicies` field.
+void Demo() {
+  // These are the values set
+  ROUTER_CUSTOM_IKEv2_POLICY0 policy;
+  policy.dwIntegrityMethod = IKEEXT_INTEGRITY_SHA_256;
+  policy.dwEncryptionMethod = IKEEXT_CIPHER_AES_GCM_256_16ICV;
+  policy.dwCipherTransformConstant = IKEEXT_CIPHER_AES_GCM_256_16ICV;
+  policy.dwAuthTransformConstant = IPSEC_CIPHER_CONFIG_GCM_AES_256;
+  policy.dwPfsGroup = IPSEC_PFS_NONE;
+  policy.dwDhGroup = IKEEXT_DH_ECP_384;
+
+  // This is the byte order they are in `CustomIPSecPolicies` field
+  // inside `%APPDATA%\Microsoft\Network\Connections\Pbk\rasphone.pbk`
+  BYTE custom_ipsec_policies[24] = {0};
+  copy_dword_bytes(&custom_ipsec_policies[0], policy.dwIntegrityMethod);
+  copy_dword_bytes(&custom_ipsec_policies[4], policy.dwEncryptionMethod);
+  copy_dword_bytes(&custom_ipsec_policies[8], policy.dwCipherTransformConstant);
+  copy_dword_bytes(&custom_ipsec_policies[12], policy.dwAuthTransformConstant);
+  copy_dword_bytes(&custom_ipsec_policies[16], policy.dwDhGroup);
+  copy_dword_bytes(&custom_ipsec_policies[20], policy.dwPfsGroup);
+
+  // characters are either written in 02d or 02x format
+  wprintf(L"\nDEMO:\n");
+  for (DWORD i = 0; i < 24; i++) {
+    wprintf(L"%02d", custom_ipsec_policies[i]);
+  }
+  wprintf(L"\n");
+}
+
+// Test program for create/remove Windows VPN entry.
+int main(int argc, char* argv[]) {
+  base::CommandLine::Init(argc, argv);
+
+  if (base::CommandLine::ForCurrentProcess()->HasSwitch(kConnectionsCommand)) {
+    PrintConnections();
+    return 0;
```

Comment:
Returning immediately would make it so that you can't run multiple commands at once. I think this is OK; but you may want to get entries and connections in the same usage